File: 3-HARD/Python/8-regular-expression-matching-dp/proposed-solution/main.py
# Challenge: Design a regular expression matching engine that supports special characters like '.' and '*',
# using dynamic programming techniques to efficiently handle branching and backtracking.
# Use object-oriented programming and follow the DRY principle.

from regex_matcher import RegexMatcher

def main():
    text = "aab"
    pattern = "c*a*b"
    matcher = RegexMatcher(text, pattern)
    result = matcher.is_match()
    print(f"Text: {text}")
    print(f"Pattern: {pattern}")
    print(f"Match result: {result}")

if __name__ == "__main__":
    main()

File: 3-HARD/JavaScript/8-regular-expression-matching-dp/proposed-solution/main.js
/*
Challenge:
Design a regular expression matching engine that supports special characters like '.' and '*', using dynamic programming techniques to efficiently handle branching and backtracking.

This solution follows DRY principles and is implemented in JavaScript.
*/

import RegexMatcher from "./RegexMatcher.js";

function main() {
  const text = "aab";
  const pattern = "c*a*b";

  const matcher = new RegexMatcher(text, pattern);
  const result = matcher.isMatch();

  console.log(
    `Does the text "${text}" match the pattern "${pattern}"?`,
    result
  );
}

main();
